source/features/action-pr-link.tsx

```diff
@@ -2,25 +2,44 @@ import * as pageDetect from 'github-url-detection';
 
 import features from '../feature-manager.js';
 import observe from '../helpers/selector-observer.js';
+import GitHubURL from '../github-helpers/github-url.js';
 
-async function add(prLink: HTMLAnchorElement): Promise<void> {
+function setSearchParameter(anchorElement: HTMLAnchorElement, name: string, value: string): void {
+	const parameters = new URLSearchParams(anchorElement.search);
+	parameters.set(name, value);
+	anchorElement.search = String(parameters);
+}
+
+async function addForRepositoryActions(prLink: HTMLAnchorElement): Promise<void> {
 	const prNumber = prLink.textContent!.slice(1);
 
 	const runLink = prLink.closest('.Box-row')!.querySelector('a.Link--primary')!;
-	const parameters = new URLSearchParams(runLink.search);
-	parameters.set('pr', prNumber);
-	runLink.search = String(parameters);
+	setSearchParameter(runLink, 'pr', prNumber);
 }
 
-async function init(signal: AbortSignal): Promise<void> {
-	observe('div.Box-row[id^=check_suite_] a[data-hovercard-type="pull_request"]', add, {signal});
+async function addForPR(actionLink: HTMLAnchorElement): Promise<void> {
+	const {branch: prNumber} = new GitHubURL(location.href);
+	setSearchParameter(actionLink, 'pr', prNumber);
```

Comment:
`GitHubURL` is only for for files. Use `getConversationNumber` instead

Reply:
I think the utilities are little bit huddle for new contributors.
It will be better when the utilities are scoped like `GitHubURL.of(location.href) returns GitHubConversationURL`. and it will make better intellisense. But this is just my opinion as user of strongly typed language like Kotlin and C#.

Reply:
> GitHubURL.of(location.href) returns GitHubConversationURL

This isn't possible in TypeScript. The returned type depends on a runtime string so the compile-time type would be `GitHubFileURL | GitHubConversationURL`, which isn't particularly useful.

`GitHubURL` exists because extracting the branch information from URLs can become complex, for example in case of branches that looks like `feature/something-something`. Extracting the number is instead very easy, it's basically a `href.split('/')[4]` and some checks.

I agree that `GitHubURL` is too generically-named so it should probably be renamed to `GitHubFile` (separate PR welcome)

Reply:
Ooops, sorry for the confusion.
More typescript-ly, I suggested like this one.

```typescript
// Like this
const features = {
	add,
	unload,
	addCssFeature,
	log,
	shortcutMap,
	getFeatureID,
	getIdentifiers,
};

export default features;

// Suggestion (The naming is too java style, but the structure)
function ofFile(): GitHubFile { ... }
function ofConversation(): GitHubConversation { ... }

const GitHubURL = {
    ofFile,
    ofConversation, // ofPullRequest | ofIssue
}

export default GitHubURLs;
```

But I think this is not generic in typescript ecosystem, and it could be because I am not familiar with it.
Never mind! Thank you for the detailed explanation.

Reply:
> Also, my suggestion seems to require an over-implementation

Yeah, GitHubURL returns a lot of information for file paths, but:

- other paths are simpler, don't need such specific parsing, or are just not parsed often enough
- GitHub has a _lot_ of different URLs, as can be seen on https://refined-github.github.io/github-url-detection/

While it'd be nice to have a consistent API, it's definitely not worth the effort